sklearn/tree/_splitter.pyx

```diff
@@ -42,6 +43,155 @@ cdef float32_t FEATURE_THRESHOLD = 1e-7
 # in SparsePartitioner
 cdef float32_t EXTRACT_NNZ_SWITCH = 0.1
 
+
+cdef bint min_sample_leaf_condition(
+    Splitter splitter,
+    SplitRecord* current_split,
+    intp_t n_missing,
+    bint missing_go_to_left,
+    float64_t lower_bound,
+    float64_t upper_bound,
+    SplitConditionEnv split_condition_env
+) noexcept nogil:
+    cdef intp_t min_samples_leaf = splitter.min_samples_leaf
+    cdef intp_t end_non_missing = splitter.end - n_missing
+    cdef intp_t n_left, n_right
+
+    if missing_go_to_left:
+        n_left = current_split.pos - splitter.start + n_missing
+        n_right = end_non_missing - current_split.pos
+    else:
+        n_left = current_split.pos - splitter.start
+        n_right = end_non_missing - current_split.pos + n_missing
+
+    # Reject if min_samples_leaf is not guaranteed
+    if n_left < min_samples_leaf or n_right < min_samples_leaf:
+        return False
+
+    return True
+
+cdef class MinSamplesLeafCondition(SplitCondition):
+    def __cinit__(self):
+        self.c.f = min_sample_leaf_condition
+        self.c.e = NULL # min_samples is stored in splitter, which is already passed to f
+
+cdef bint min_weight_leaf_condition(
+    Splitter splitter,
+    SplitRecord* current_split,
+    intp_t n_missing,
+    bint missing_go_to_left,
+    float64_t lower_bound,
+    float64_t upper_bound,
```

Comment:
If I read the existing code correctly, those values are set in `current_split` only _after_ it is accepted as passing pre-and-post split conditions, and yielding a greater impurity improvement than `best_split`, so that the `n_missing`, `missing_go_to_left`, `lower_bound`, and `upper_bound` values in `current_split` are potentially garbage values at the time these split rejection conditions are tested.

Reply:
Perhaps a better alternative to passing `current_split` into these split rejection conditions would be to simply pass the candidate feature dimension and split point.

Reply:
Yeah I would say either:

i) pass in only SplitRecord and let the function implementation worry about what is garbage vs not, cuz you shouldn't use garbage values anyways
ii) only pass in parameters that are necessary

It shouldn't pass in split record and parameters explicitly.

Reply:
What it will likely ultimately end up looking like is something similar to the BuildEnv struct pattern added to `DepthFirstTreeBuilder.build`... I haven't yet thought hard about what the final form of this signature should look like, but it would need to contain all the splitter state information necessary for arbitrary split constraints to decide thumbs up or down. The platonic ideal would make that part of a deliberately curated interface.